File: DotNetPractices/ProblemSolving/c/quickSort.c

void swap(int *a, int *b) {
    int t = *a;
    *a = *b;
    *b = t;
}

int partition(int a[], int low, int high) {
    int pivot = a[high];
    int l = low;

    for(int i=low;i<high-1;i++) {
        if(a[i]<pivot) {
            swap(&a[i], &a[l]);
            l++;
        }
    }
    swap(&a[l], &a[high]);
    return l;
}


void quickSort(int a[], int low, int high) {
    while(low<high) {
        int pi = partition(a, low, high);

        quickSort(a, low, pi-1);
        quickSort(a, pi+1, high);
    }
}

File: DotNetPractices/ProblemSolving/Sorting/QuickSort.cs
namespace Programming.Sorting
{
    internal class QuickSort
    {
        public int[] Solution(int[] nums)
        {
            Sort(nums, 0, nums.Length - 1);

            return nums;
        }


        private void Sort(int[] nums, int low, int high)
        {

            if (low > high) return;

            int pi = Partition(nums, low, high);

            Sort(nums, low, pi - 1);
            Sort(nums, pi + 1, high);
        }

        private int Partition(int[] nums, int low, int high)
        {
            var pivot = nums[high];

            int i = low - 1;
            for (int j = low; j < high; j++)
            {
                if (nums[j] < pivot)
                {
                    i++;
                    Swap(nums, i, j);
                }
            }

            Swap(nums, i + 1, high);

            return i + 1;
        }


        public void Swap(int[] nums, int i, int j)
        {
            (nums[j], nums[i]) = (nums[i], nums[j]);
        }
    }


}
